File: src/car_base/test/test_protocol.py
"""协议编解码单元测试。

用《总体方案设计 v1.2.1》6.5 节给出的确切示例值做断言。
无需硬件,WSL 上即可运行:  python3 -m pytest test/test_protocol.py -v
"""

import struct
import pytest

from car_base.protocol import (
    build_ctrl_frame, parse_feedback, bcc, FrameReader,
    FRAME_HEAD, FRAME_TAIL, CTRL_FRAME_LEN, FB_FRAME_LEN,
)


def make_fb_frame(vx_raw, vy_raw, vz_raw, ax, ay, az, gx, gy, gz, batt_mv):
    """按协议拼一个合法 24 字节反馈帧(自动算 BCC)。"""
    body = struct.pack('>BB hhh hhh hhh H',
                       FRAME_HEAD, 0,
                       vx_raw, vy_raw, vz_raw,
                       ax, ay, az,
                       gx, gy, gz,
                       batt_mv)
    assert len(body) == 22
    return body + bytes([bcc(body), FRAME_TAIL])


def test_ctrl_frame_structure():
    f = build_ctrl_frame(0.257, 0.0)
    assert len(f) == CTRL_FRAME_LEN
    assert f[0] == FRAME_HEAD
    assert f[-1] == FRAME_TAIL
    # BCC 覆盖字节 1..9(索引 0..8)
    assert f[9] == bcc(f[0:9])
    # vx = 0.257 m/s -> 257 = 0x0101 大端
    assert f[2] == 0x01 and f[3] == 0x01


def test_ctrl_frame_negative_vx():
    # -0.257 m/s -> -257 补码大端
    f = build_ctrl_frame(-0.257, 0.0)
    vx_raw = struct.unpack('>h', f[2:4])[0]
    assert vx_raw == -257


def test_ctrl_wz_sign():
    f_pos = build_ctrl_frame(0.0, 0.5, cmd_wz_sign=1)
    f_neg = build_ctrl_frame(0.0, 0.5, cmd_wz_sign=-1)
    wz_pos = struct.unpack('>h', f_pos[6:8])[0]
    wz_neg = struct.unpack('>h', f_neg[6:8])[0]
    assert wz_pos == 500
    assert wz_neg == -500


def test_parse_feedback_doc_example():
    # 文档 6.5: X速度 0x0101=257 -> 0.257 m/s
    #          Z加速度 0x4080=16512 -> /1672 = 9.8756 m/s^2
    #          电池 0x5838=22584 mV -> 22.584 V
    frame = make_fb_frame(
        vx_raw=0x0101, vy_raw=0, vz_raw=0,
        ax=0, ay=0, az=0x4080,
        gx=0, gy=0, gz=0,
        batt_mv=0x5838)
    fb = parse_feedback(frame)
    assert fb.vx == pytest.approx(0.257, abs=1e-6)
    assert fb.acc[2] == pytest.approx(9.8756, abs=1e-3)
    assert fb.voltage == pytest.approx(22.584, abs=1e-6)


def test_parse_feedback_negative():
    # 文档 6.5 负数示例: 0xFE96=-362 -> /1672=-0.2165 m/s^2
    #                    0xFFFB=-5 -> /3753=-0.00133 rad/s
    frame = make_fb_frame(
        vx_raw=0, vy_raw=0, vz_raw=0,
        ax=struct.unpack('>h', b'\xFE\x96')[0], ay=0, az=0,
        gx=0, gy=0, gz=struct.unpack('>h', b'\xFF\xFB')[0],
        batt_mv=12000)
    fb = parse_feedback(frame)
    assert fb.acc[0] == pytest.approx(-0.2165, abs=1e-3)
    assert fb.gyro[2] == pytest.approx(-0.00133, abs=1e-4)


def test_bcc_detects_corruption():
    frame = bytearray(make_fb_frame(1, 0, 0, 0, 0, 0, 0, 0, 0, 12000))
    frame[5] ^= 0xFF                    # 破坏一个数据字节
    with pytest.raises(ValueError, match='BCC'):
        parse_feedback(bytes(frame))


def test_frame_reader_resync_with_garbage():
    """流里混入垃圾字节 + 半帧,验证帧同步能恢复。"""
    good = make_fb_frame(0x0101, 0, 0, 0, 0, 0, 0, 0, 0, 12000)
    reader = FrameReader()
    # 前面垃圾 + 一整帧 + 下一帧的前半段
    stream = b'\x00\xFF\xAB' + good + good[:10]
    out = reader.feed(stream)
    assert len(out) == 1
    assert out[0].vx == pytest.approx(0.257, abs=1e-6)
    # 再喂入后半段,应拼出第二帧
    out2 = reader.feed(good[10:])
    assert len(out2) == 1
    assert reader.ok == 2


def test_roundtrip_wz_sign():
    """发送套 cmd_wz_sign,接收套 odom_wz_sign,同号应还原。"""
    sign = -1
    f = build_ctrl_frame(0.0, 0.4, cmd_wz_sign=sign)
    raw = struct.unpack('>h', f[6:8])[0]      # = -400
    # 接收端用同样符号还原
    wz_ros = sign * raw / 1000.0
    assert wz_ros == pytest.approx(0.4, abs=1e-6)

File: src/car_base/test/test_protocol.cpp
// C++ 协议编解码单元测试,用《总体方案设计 v1.2.1》6.5 节示例值断言。
#include <gtest/gtest.h>
#include <array>
#include <cstring>
#include <vector>

#include "car_base/protocol.hpp"

using namespace car_base;

// 按协议拼一个合法 24 字节反馈帧(自动算 BCC)
static std::array<uint8_t, kFbFrameLen> make_fb(
  int16_t vx, int16_t vy, int16_t vz,
  int16_t ax, int16_t ay, int16_t az,
  int16_t gx, int16_t gy, int16_t gz,
  uint16_t batt_mv)
{
  std::array<uint8_t, kFbFrameLen> f{};
  f[0] = kFrameHead;
  f[1] = 0;
  write_be_i16(&f[2], vx);
  write_be_i16(&f[4], vy);
  write_be_i16(&f[6], vz);
  write_be_i16(&f[8], ax);
  write_be_i16(&f[10], ay);
  write_be_i16(&f[12], az);
  write_be_i16(&f[14], gx);
  write_be_i16(&f[16], gy);
  write_be_i16(&f[18], gz);
  f[20] = static_cast<uint8_t>((batt_mv >> 8) & 0xFF);
  f[21] = static_cast<uint8_t>(batt_mv & 0xFF);
  f[22] = bcc(f.data(), 22);
  f[23] = kFrameTail;
  return f;
}

TEST(Protocol, CtrlFrameStructure) {
  auto f = build_ctrl_frame(0.257, 0.0);
  EXPECT_EQ(f.size(), kCtrlFrameLen);
  EXPECT_EQ(f[0], kFrameHead);
  EXPECT_EQ(f[10], kFrameTail);
  EXPECT_EQ(f[9], bcc(f.data(), 9));
  // 0.257 m/s -> 257 = 0x0101 大端
  EXPECT_EQ(f[2], 0x01);
  EXPECT_EQ(f[3], 0x01);
}

TEST(Protocol, CtrlFrameNegative) {
  auto f = build_ctrl_frame(-0.257, 0.0);
  EXPECT_EQ(read_be_i16(&f[2]), -257);
}

TEST(Protocol, CtrlWzSign) {
  auto fp = build_ctrl_frame(0.0, 0.5, 1);
  auto fn = build_ctrl_frame(0.0, 0.5, -1);
  EXPECT_EQ(read_be_i16(&fp[6]), 500);
  EXPECT_EQ(read_be_i16(&fn[6]), -500);
}

TEST(Protocol, CtrlVxSign) {
  // 第4个参数 cmd_vx_sign:-1 应把正 vx 编码成负值
  auto fp = build_ctrl_frame(0.2, 0.0, 1, 1);
  auto fn = build_ctrl_frame(0.2, 0.0, 1, -1);
  EXPECT_EQ(read_be_i16(&fp[2]), 200);
  EXPECT_EQ(read_be_i16(&fn[2]), -200);
}

TEST(Protocol, ParseFeedbackDocExample) {
  // X 速度 0x0101=257 -> 0.257 m/s;Z 加速度 0x4080=16512 -> /1672=9.8756;
  // 电池 0x5838=22584 mV
  auto f = make_fb(0x0101, 0, 0, 0, 0, 0x4080, 0, 0, 0, 0x5838);
  auto fb = parse_feedback(f.data(), f.size());
  ASSERT_TRUE(fb.has_value());
  EXPECT_NEAR(fb->vx, 0.257, 1e-6);
  EXPECT_NEAR(fb->acc[2], 9.8756, 1e-3);
  EXPECT_NEAR(fb->voltage_raw, 22.584, 1e-6);
}

TEST(Protocol, ParseFeedbackNegative) {
  // 0xFE96=-362 -> /1672=-0.2165;0xFFFB=-5 -> /3753=-0.00133
  auto f = make_fb(0, 0, 0,
    static_cast<int16_t>(0xFE96), 0, 0,
    0, 0, static_cast<int16_t>(0xFFFB), 12000);
  auto fb = parse_feedback(f.data(), f.size());
  ASSERT_TRUE(fb.has_value());
  EXPECT_NEAR(fb->acc[0], -0.2165, 1e-3);
  EXPECT_NEAR(fb->gyro[2], -0.00133, 1e-4);
}

TEST(Protocol, BccDetectsCorruption) {
  auto f = make_fb(1, 0, 0, 0, 0, 0, 0, 0, 0, 12000);
  f[5] ^= 0xFF;  // 破坏数据字节
  auto fb = parse_feedback(f.data(), f.size());
  EXPECT_FALSE(fb.has_value());
}

TEST(Protocol, FrameReaderResyncWithGarbage) {
  auto good = make_fb(0x0101, 0, 0, 0, 0, 0, 0, 0, 0, 12000);
  FrameReader reader;
  // 垃圾字节 + 一整帧
  std::vector<uint8_t> stream = {0x00, 0xFF, 0xAB};
  stream.insert(stream.end(), good.begin(), good.end());
  // 加下一帧前 10 字节
  stream.insert(stream.end(), good.begin(), good.begin() + 10);
  auto out = reader.feed(stream.data(), stream.size());
  ASSERT_EQ(out.size(), 1u);
  EXPECT_NEAR(out[0].vx, 0.257, 1e-6);
  // 喂入后半段,拼出第二帧
  auto out2 = reader.feed(good.data() + 10, kFbFrameLen - 10);
  ASSERT_EQ(out2.size(), 1u);
  EXPECT_EQ(reader.ok, 2u);
}

int main(int argc, char ** argv)
{
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
